### DISCUSSION TITLE
some json validation can't provide the detail violation path

### DISCUSSION THREAD

**Discussion by yimingming** [danielaparker/jsoncons#364] [Q&A] [answered]
Hi experts,
I use jsoncons to test json validation. it seems that it can't  provide the detail violation path in some case, for example, when the schema using allOf, anyOf, etc.  Could you please help me to solve the issue? I hope to get the complete violation path just like "Schema path: #/allOf/1/then/properties/abcInfo/required" and "Message:Required properties are missing from object: suRanges." in the following case . Thanks a lot.

source code:
#include <jsoncons/json.hpp>
#include <jsoncons/json_exception.hpp>
#include <jsoncons_ext/jsonschema/jsonschema.hpp>
using namespace jsoncons;
using namespace jsonschema;

int main() 
{
    json schema = json::parse(R"(
    {"type": "object",
    "properties":
    {
        "nfType":
        {
            "type": "string"
        }
    },
    "allOf": [
        {
            "description": "at least one of (iRanges/pList or xRanges) required for each xyzinfo",
            "if":
            {
                "properties":
                {
                    "nfType":
                    {
                        "const": "XYZ"
                    }
                }
            },
            "then":
            {
                "properties":
                {
                    "xyzInfo":
                    {
                        "type": "object",
                        "anyOf": [
                            {
                                "required": ["iRanges", "pList"]
                            },
                            {
                                "required": ["xRanges"]
                            }
                        ]
                    }
                }
            }
        },
        {
            "description": "su/si required for each abcinfo",
            "if":
            {
                "properties":
                {
                    "nfType":
                    {
                        "const": "ABC"
                    }
                }
            },
            "then":
            {
                "properties":
                {
                    "abcInfo":
                    {
                        "required": ["suRanges", "siRanges"]
                    }
                }
            }
        }
    ]})");

    json data = json::parse(R"(
    {"nfId":"ycudm0000001",
    "nfType":"ABC",
    "nfStatus":"REGISTERED",
    "abcInfo":
      {"groupId":"A1001",
       "siRanges":[
            {"start":"8613300001","end":"8613309999"},
            {"start":"8613310001","end":"8613319999"}
                  ]
      },
    "recoveryTime":"2019-10-28t13:27:43z"
    })");
    
    try
    {
        auto sch = jsonschema::make_schema(schema);
        std::size_t error_count = 0;
        auto reporter = [&error_count](const jsonschema::validation_output& o)
        {
           ++error_count;
           std::cout<<"validation error : " << o.instance_location() << ": " << o.message() << "\n";
        };
           jsonschema::json_validator<json> validator(sch);
           validator.validate(data, reporter);
           std::cout << "\nError count: " << error_count << "\n\n";
    }
    catch (const std::exception& e)
    {
        std::cout<<"exception : " << e.what() << "\n";
    }
    return 0;
}

output:
validation error : #: At least one keyword_validator failed to match, but all are required to match. 

Error count: 1

**Comment by danielaparker** [comment, marked as answer]
The detail is in `nested_errors()`, if you change your reporter to this
```cpp
        auto reporter = [&error_count](const jsonschema::validation_output& o)
        {
            ++error_count;
            std::cout << "validation error : " << o.instance_location() << ": " << o.message() << "\n";
            for (auto& err : o.nested_errors())
            {
                std::cout << "    " << err.instance_location() << ": " << err.message() << "\n";
            }
        };
```
your output will be
```
validation error : #: At least one keyword_validator failed to match, but all are required to match.
    #/abcInfo: Required property "suRanges" not found
```

**Reply by yimingming** [reply]
Danielaparker,
Thanks a lot for your quick response and perfect solution.
